pandas/core/generic.py

```diff
@@ -1940,6 +1940,76 @@ def tail(self, n=5):
             return self
         return self.iloc[-n:]
 
+  
+    def sample(self, n=None, frac=None , replace=False, weights=None, seed=None):
+        """
+        Returns a random sample of rows from object. 
+        
+        Parameters
+        ----------
+        n: int, optional.
+            Number of rows to return. Cannot be used with frac.
+            Default = 5 if frac = None. 
+        frac: float. 
+            Share of rows to return. Cannot be used with n. 
+            Default = None.
+        replace: boolean. 
+            Sample with or without replacement. Default = False. 
+        weights: str or ndarray-like, optional.
+            Default 'None' results in equal probability weighting. 
+            If called on a DataFrame or Panel, will also accept the name of a 
+            column as a string. Must be same length as index. 
+            Weights must be normalized. 
+        seed: int.
+            Seed for the random number generator.  Default = None. 
+            
+        Returns
+        -------
+        Same type as caller. 
+
+        -----------
+        """
+    
+        # Setup vars
+        rs = np.random.RandomState(seed)
+    
+        # Check weights for compliance
+        if weights is not None:
+                                      
+            # Strings acceptable if not a series
+            if isinstance(weights, string_types) and self.ndim >1 :
+                try:
+                    weights = self[weights]
+                except(KeyError):
+                    raise KeyError("String passed to weights not a valid column name")
+            
+            if isinstance(weights, string_types) and self.ndim == 1:
```

Comment:
rather than testing the type of `weights` twice, do these checks within a single if statement.